[repo] docs: added concepts to the getting started docs (#1176)
microsoft/teams-ai: getting-started/CONCEPTS/PROMPTS.md

# Prompts

Prompts play a crucial role in communicating and directing the behavior of Large Language Models (LLMs) AI. 
They serve as inputs or queries that users can provide to elicit specific responses from a model.

Here's a prompt that asks the LLM for name suggestions:

_Input:_

```prompt
Give me 3 name suggestions for my pet golden retriever.
```

_Response:_

> ```prompt

Reviewer: Is prompt what we want to put here?

Some possible name suggestions for a pet golden retriever are:

- Bailey
- Sunny
- Cooper
```

# Prompt Template

Prompt templates are a simple and powerful way to
define and compose AI functions **using plain text**.
You can use it to create natural language prompts, generate responses, extract

> information, **invoke other prompts** or perform any other task that can be

Reviewer: ```suggestion
information, **invoke other prompts,** or perform any other task that can be
```

expressed with text.

The language supports two basic features that allow you to include
variables and call functions.

**Simple Example:**

Here's an example of a prompt template:

```
Give me 3 name suggestions for my pet {{ $petName }}.
```

`$petName` is a variable that is populated on runtime when the template is rendered.

## Prompt Template Language

You don't need to write any code or import any external libraries, just use the

> curly braces {{...}} to embed expressions in your prompts.

Reviewer: ```suggestion
double curly braces {{...}} to embed expressions in your prompts.
```

Teams AI will parse your template and execute the logic behind it.
This way, you can easily integrate AI into your apps with minimal effort and
maximum flexibility.

### Variables

To include a variable value in your text, use the `{{$variableName}}` syntax. For example, if you have a variable called name that holds the user's name, you can write:

`Hello {{$name}}, nice to meet you!`

This will produce a greeting with the user's name.

> Spaces are ignored, so if you find it more readable, you can also write:

Reviewer: Up to you , but I'm not sure we want to claim this quite yet because of the following bug: https://github.com/microsoft/teams-ai/issues/1087

Author: Let's fix that bug soon.

`Hello {{ $name }}, nice to meet you!`

Here's how to define variables in code:

**C#**

In an *action* or *route handler* where the turn state object is available:
```cs

> state.Temp.Post = "Lorem Ipsium..."

Reviewer: ```suggestion
state.Temp.Post = "Lorem Ipsum..."
```

```

The usage in the prompt:

> ```

Reviewer: ```suggestion
```prompt
```

Author: just ``` should be fine since it's a text block

This is the user's post: {{ $post }}
```

> > Note: The `turnState.Temp.Post = ...` updates a dictionary with the `post` key under the hood from the [GPT Message Extension sample](https://github.com/microsoft/teams-ai/blob/a20f8715d3fe81e11c330853e3930e22abe298af/dotnet/samples/04.ai.b.messageExtensions.gptME/ActivityHandlers.cs#L156).

Reviewer: ```suggestion
> Note: The `turnState.Temp.Post = ...` updates a dictionary with the `post` key under the hood from the [Message Extension sample](https://github.com/microsoft/teams-ai/blob/a20f8715d3fe81e11c330853e3930e22abe298af/dotnet/samples/04.ai.b.messageExtensions.gptME/ActivityHandlers.cs#L156).
```
We're supposed to avoid using GPT since it's proprietary, I believe. Unless there has been a change?

Reviewer: Should also link to the JS sample

**Javascript**

```typescript
app.beforeTurn((context, state) => {

>   state.temp.post = "Lorem Ipsium...";

Reviewer: ```suggestion
  state.temp.post = "Lorem Ipsum...";
```

});
```

The usage in the prompt:

> ```

Reviewer: ```suggestion
```prompt
```

This is the user's post: {{ $post }}
```

You can simply add to the `state.temp` object, and it will be accessible from the prompt template on runtime. Note that the safest place to do that would be in the `beforeTurn` activity because it will execute before any activity handler or action.

**Default Variables**

> The following are variables accesible in the prompt template without having to manually configure. These are pre-defined in the turn state and populated by the library. Users can override them by changing it in the turn state.

Reviewer: ```suggestion
The following are variables accessible in the prompt template without having to manually configure them. These are pre-defined in the turn state and populated by the library. Users can override them by changing it in the turn state.
```

| Variable name | Description                                                       |
| ------------- | ----------------------------------------------------------------- |
| `input`       | Input passed from the user to the AI Library. |

> | `lastOutput`      | Output returned from the last executed action.            |

Reviewer: | Variable name | Description                                    |
| ------------- | ---------------------------------------------- |
| `input`       | Input passed from the user to the AI Library.  |
| `lastOutput`  | Output returned from the last executed action. |

> | Variable name | Description                                                       |
> | ------------- | ----------------------------------------------------------------- |
> | `input`       | Input passed from the user to the AI Library. |
> | `lastOutput`      | Output returned from the last executed action.            |

Reviewer: ```suggestion
| Variable name | Description                                    |
| ------------- | ---------------------------------------------- |
| `input`       | Input passed from the user to the AI Library.  |
| `lastOutput`  | Output returned from the last executed action. |
```

### Function calls

> To call an external function and embed the result in your text, use the `{{ functionName }}` syntax. For example if you have a function called `diceRoll` that returns a random number between 1 and 6, you can write:

Reviewer: ```suggestion
To call an external function and embed the result in your text, use the `{{ functionName }}` syntax. For example, if you have a function called `diceRoll` that returns a random number between 1 and 6, you can write:
```

`The dice roll has landed on: {{ diceRoll }}`

**C#**

In the `Application` class,

```cs
prompts.AddFunction("diceRoll", async (context, memory, functions, tokenizer, args) =>
{
    int diceRoll = // random number between 1 and 6
    return diceRoll;
});
```

**Javascript**

```typescript
prompts.addFunction('diceRoll', async (context, state, functions, tokenizer, args) => {
    let diceRoll = // random number between 1 and 6
    return diceRoll;
});
```

# Creating Prompt Templates

> Each prompt template is a folder with two files, `skprompt.txt` and `config.json`. The folder name is the prompt template's name that can be referred in code. The `skprompt.txt` file contains the prompt's text, which can contain natural language or prompt template syntax as defined in the previous section. The `config.json` file specifies the prompt completion configuration.

Reviewer: ```suggestion
Each prompt template is a folder with two files, `skprompt.txt` and `config.json`. The folder name is the prompt template's name which can be referred to in your code. The `skprompt.txt` file contains the prompt's text, which can contain natural language or prompt template syntax as defined in the previous section. The `config.json` file specifies the prompt completion configuration.
```

Here's an example of a prompt template from the [Twenty Questions](https://github.com/microsoft/teams-ai/blob/c5ec11842b808e48cd214b3cb52da84e5811da33/js/samples/04.e.twentyQuestions) sample.

*skprompt.txt*
```
You are the AI in a game of 20 questions.
The goal of the game is for the Human to guess a secret within 20 questions.
The AI should answer questions about the secret.
The AI should assume that every message from the Human is a question about the secret.

GuessCount: {{$conversation.guessCount}}
RemainingGuesses: {{$conversation.remainingGuesses}}
Secret: {{$conversation.secretWord}}

> Answer the humans question but do not mention the secret word.

Reviewer: ```suggestion
Answer the human's question but do not mention the secret word.
```

```

*config.json*
```json
{
    "schema": 1.1,
    "description": "A bot that plays a game of 20 questions",
    "type": "completion",
    "completion": {
        "completion_type": "chat",
        "include_history": false,
        "include_input": true,
        "max_input_tokens": 2000,
        "max_tokens": 256,
        "temperature": 0.7,
        "top_p": 0.0,
        "presence_penalty": 0.6,
        "frequency_penalty": 0.0
    }
}
```

> Note that the configuration properties in the file do not include all the possible configurations. To learn more about the description of each configuration and all the supported configurations see the [`PromptTemplatConfig`](https://github.com/microsoft/teams-ai/blob/2d43f5ca5b3bf27844f760663641741cae4a3243/js/packages/teams-ai/src/prompts/PromptTemplate.ts#L46C18-L46C39) Typescript interface. 

> These files can be found under the `src/prompts/chat/` folder. So this prompt template's name is `chat`. Then to plug these files in the Action Planner, the prompt manager has to be created with the folder path specified and then passed into the Action Planner constructor:

Reviewer: ```suggestion
These files can be found under the `src/prompts/chat/` folder. So, this prompt template's name is `chat`. Then, to plug these files in the Action Planner, the prompt manager has to be created with the folder path specified and then passed into the Action Planner constructor:
```
